packages/yarnpkg-core/sources/Controllers.ts

```diff
@@ -0,0 +1,105 @@
+import {Fetcher}                                         from './Fetcher';
+import {MultiFetcher}                                    from './MultiFetcher';
+import {MultiResolver}                                   from './MultiResolver';
+import {Resolver, ResolveOptions, MinimalResolveOptions} from './Resolver';
+import {Descriptor, Locator, DescriptorHash, Package}    from './types';
+
+// The CoreResolver and CoreFetcher classes are controllers for the actual resolvers/fetchers.
+// This description uses a resolver for example, but the logic is the same for fetchers.
+// All resolvers with type 'module' are stored in an array `resolvers.module`
+// All resolvers with anything else (ie. type: 'protocol') are loaded into a MultiResolver
+// and stored in `resolvers.protocol`.
+// The logic is that the first loaded module is consulted first with the following options
+// {next: nextResolver, proto: resolvers.proto}. The resolver can choose to return a response,
+// or consult the next resolver in the chain, or consult the protocol resolvers.
+// `next` is a wrapper function to call the next resolver passing it the resolver after that
+
+export class ResolverController implements Resolver {
+  public readonly type = 'module';
+  private readonly index = new Map<Resolver, Resolver | undefined>(); // map of each module resolver to the next in the chain
+  private readonly protocols = new MultiResolver([]); // a MultiResolver aggregating all protocol resolvers
+  private readonly modules = new Array<Resolver>(); // an array of all modules (mostly to keep track of which to call first)
+
+  constructor(resolvers: Array<Resolver>) {
+    resolvers.forEach( (resolver) => { // categorize each plugin
+      if (resolver && resolver.type === 'module') {
+        this.modules.push(resolver);
+      } else if (resolver) {
+        this.protocols.addResolver(resolver);
+      }
+    } );
+    for (let i=0; i < this.modules.length; i++) { // create a map of each module with its next (ie. the chain of module resolvers)
+      this.index.set(this.modules[i], this.modules[i+1]);
+    }
+  }
+
+  prepareNext(func: string, resolver: Resolver | undefined, ...args: any): any {
+    if (resolver) {
+      const next = this.index.get(resolver); // if next is undefined the `prepareNext` will use the protocol MultiResolver
+      return (resolver as any)[func].bind(resolver, ...args, this.prepareNext(func, next, ...args), this.protocols);
+    } else {
+      return (this.protocols as any)[func].bind(this.protocols, ...args);
+    }
+  }
+
+  async supportsDescriptor(descriptor: Descriptor, opts: MinimalResolveOptions): Promise<boolean> {
+    return await this.prepareNext('supportsDescriptor', this.modules[0], descriptor, opts)();
```

Comment:
Not a fan of using one method for everything - at the very least `prepareNext` should be a generic and its first argument should be a `keyof Resolver`.

Reply:
`prepareNext` is just a wrapper function that binds the resolver function to the arguments and adds to too additional arguments next and protocol. The first argument is the function that should be bound (string), the second argument is the object from which to get the function to be bound (the resolver being wrapped), and the remaining spread operators are the arguments that are actually passed to the controller (passed through and bound to the resolver function). It probably should be generic, but I am new to typescript and forgot about generics.

In retrospect, it does seem that it would be a lot simpler to just pass the next resolver into the constructor of the current resolver (i guess they are initialized in reverse order). I assume that all of the existing resolvers would still be wrapped up into a single MultiResolver, so we are really mostly talking about new plugins people write in the future.